### PR TITLE
chore(v2): support changing pipeline root

### LINKED ISSUE #6153: v2 placeholders - artifact passing via URI

### PR DESCRIPTION
**Pull request by Bobgy**
**Description of your changes:**
Fixes #6153 

**Checklist:**
- [ ] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: v2/compiler/argo.go [hunk 1]
```diff
@@ -68,6 +76,12 @@ func Compile(job *pipelinespec.PipelineJob, opts *Options) (*wfapi.Workflow, err
 		if opts.LauncherImage != "" {
 			compiler.launcherImage = opts.LauncherImage
 		}
+		if opts.PipelineRoot != "" {
+			if job.RuntimeConfig == nil {
+				job.RuntimeConfig = &pipelinespec.PipelineJob_RuntimeConfig{}
+			}
+			job.RuntimeConfig.GcsOutputDirectory = opts.PipelineRoot
```
Comment (zijianjoy) [inline v2/compiler/argo.go:83]: Can PipelineRoot be something other than GCS path? Should we set GcsOutputDirectory if pipelineRoot is minio or other platform?

File: v2/driver/driver.go [hunk 2]
```diff
@@ -73,7 +73,7 @@ func RootDAG(ctx context.Context, opts Options, mlmd *metadata.Client) (executio
 	if err != nil {
 		return nil, err
 	}
-	// TODO(Bobgy): change GCS output directory to pipeline root.
+	// TODO(v2): in pipeline spec, rename GCS output directory to pipeline root.
 	pipelineRoot := opts.RuntimeConfig.GetGcsOutputDirectory()
 	if pipelineRoot != "" {
```
Comment (capri-xiyue) [inline v2/driver/driver.go:78]: I feel confused about the logic here.
Looks like the the priority of pipeline root passed via arguments is higher than that of pipeline root defined in the configmap. It makes sense to me. But why the pipeline root passed via arguments will always be gcs?

**Comment by Bobgy** [comment]
Answering both comments together, @zijianjoy @capri-xiyue we only have the GcsOutputDirectory field in pipeline spec today. We should rename the field to PipelineRoot in the future. Let's start that conversation.

For now, the implementation will just assume GcsOutputDirectory can be minio:// or s3://.

File: v2/driver/driver.go (hunk 2 above)
Comment (Bobgy) [reply v2/driver/driver.go:78]: Yes, your understanding on priority is correct. Answered the gcs question in a common comment.

**Comment by Bobgy** [comment]
Fixed a minor test error

**Comment by Bobgy** [comment]
> Traceback (most recent call last):
  File "sample_test.py", line 158, in <module>
    fire.Fire(main)
  File "/usr/local/lib/python3.7/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/usr/local/lib/python3.7/site-packages/fire/core.py", line 471, in _Fire
    target=component.__name__)
  File "/usr/local/lib/python3.7/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "sample_test.py", line 143, in main
    run_response = run_result.wait_for_run_completion(PIPELINE_TIME_OUT)
  File "/home/prow/go/src/github.com/kubeflow/pipelines/sdk/python/kfp/_client.py", line 940, in wait_for_run_completion
    return self._client.wait_for_run_completion(self.run_id, timeout)
  File "/home/prow/go/src/github.com/kubeflow/pipelines/sdk/python/kfp/_client.py", line 1065, in wait_for_run_completion
    status = get_run_response.run.status
AttributeError: 'NoneType' object has no attribute 'status'

The error seems transient.
/retest

**Comment by zijianjoy** [comment]
Sounds good, thank you Yuan for the explanation! 
We should probably define a new field for common path usage, then gradually move the gcsOutputDirectory usage to there, and make sure Vertex also aligns with this change.

**Pull request merged**